Concentus.Oggfile/OpusOggReadStream.cs

```diff
@@ -128,37 +125,82 @@ private bool Initialize()
         {
             try
             {
-                OggContainerReader reader = new OggContainerReader(_inputStream, true);
-                if (!reader.Init())
+                var oggContainerReader = new OggContainerReader(_stream, true);
+                if (!oggContainerReader.Init())
                 {
-                    _lastError = "Could not initialize stream";
+                    LastError = "Could not initialize stream";
                     return false;
                 }
 
-                //if (!reader.FindNextStream())
-                //{
-                //    _lastError = "Could not find elementary stream";
-                //    return false;
-                //}
-                if (reader.StreamSerials.Length == 0)
+                if (oggContainerReader.StreamSerials.Length == 0)
                 {
-                    _lastError = "Initialization failed: No elementary streams found in input file";
+                    LastError = "Initialization failed: No elementary streams found in input file";
                     return false;
                 }
 
-                int streamSerial = reader.StreamSerials[0];
-                _packetProvider = reader.GetStream(streamSerial);
+                int firstStreamSerial = oggContainerReader.StreamSerials[0];
+                _packetProvider = oggContainerReader.GetStream(firstStreamSerial);
+
+                GranuleCount = _packetProvider.GetGranuleCount();
+                PageCount = _packetProvider.GetTotalPageCount();
+
                 QueueNextPacket();
 
                 return true;
             }
             catch (Exception e)
             {
-                _lastError = "Unknown initialization error: " + e.Message;
+                LastError = "Unknown initialization error: " + e.Message;
                 return false;
             }
         }
 
+        /// <summary>
+        /// Seeks the Opus stream for a valid packet at the specified offset.
+        /// </summary>
+        /// <param name="offset">The sample offset.</param>
+        public void SeekTo(long offset)
+        {
+            // Find a packet based on offset and return 1 in the callback if the packet is valid
+            var foundPacket = _packetProvider.FindPacket(offset, GetPacketLength);
+
+            // Check of the found packet is valid
+            if (foundPacket == null || foundPacket.IsEndOfStream)
+            {
+                _endOfStream = true;
+                _nextDataPacket = null;
+                return;
+            }
+
+            // Just seek to this found packet and get the previous packet (preRoll = 1)
+            _packetProvider.SeekToPacket(foundPacket, 1);
+
+            // Update the PageGranulePosition to the position from this next packet which will be retrieved by the next QueueNextPacket call
+            PageGranulePosition = _packetProvider.PeekNextPacket().PageGranulePosition;
```

Comment:
I'm guessing you will want to Reset() the decoder state after the seek so that playback will start from a clean slate

Reply:
I don't know the code from the decoder, but if you say this is needed, I can add this.

Reply:
Yep, just call Reset on the opus decoder and it should be read to start playback from the new position